File: Linked list/insertion_sort_list.py
"""Sort a linked list using insertion sort."""


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if not head: return None
        if not head.next: return head
        
        dummy = ListNode(None,head)
        current = head
        
        while current.next is not None:
            if current.next.val >= current.val:
                current = current.next     
            else: 
                temp = current.next
                runner = dummy
                current.next = current.next.next
                while runner.next.val <= temp.val:
                    runner = runner.next
                runner.next, temp.next = temp, runner.next
                
        return dummy.next   
                

File: Linked list/InsertionSortList.cpp
//Problem Link:- https://leetcode.com/problems/insertion-sort-list/

#include<bits/stdc++.h>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* start = new ListNode();
        ListNode* curr = head;
        ListNode* prev = start;
        start->next = head;
        
        while (curr){
            if (curr->next && curr->val > curr->next->val){
                while(prev->next && (prev->next->val < curr->next->val)){
                    prev = prev->next;
                }
                ListNode* temp = prev->next;
                prev->next = curr->next;
                curr->next = curr->next->next;
                prev->next->next = temp;
                prev = start;
            }
            else{
                curr = curr->next;
            }
        }
        return start->next;
    }
};